crypto/pkcs7/pkcs7_x509.c

```diff
@@ -370,6 +377,100 @@ static int write_signer_info(CBB *out, const void *arg) {
   return ret;
 }
 
+static PKCS7_SIGNER_INFO *pkcs7_add_signature(PKCS7 *p7, X509 *x509,
+                                              EVP_PKEY *pkey) {
+  PKCS7_SIGNER_INFO *si = NULL;
+  const EVP_MD *digest = NULL;
+
+  EVP_PKEY_CTX *pkey_ctx = EVP_PKEY_CTX_new(pkey, /*engine*/ NULL);
+  if (!pkey_ctx) {
+    goto err;
+  }
+  int ok = EVP_PKEY_CTX_get_signature_md(pkey_ctx, &digest);
+  EVP_PKEY_CTX_free(pkey_ctx);
+  if (!ok && (EVP_PKEY_id(pkey) == EVP_PKEY_RSA ||
+              EVP_PKEY_id(pkey) == EVP_PKEY_DSA)) {
+    // OpenSSL's docs say that this defaults to SHA1, but appears to actually
+    // default to SHA256 in 1.1.x and 3.x
+    // https://linux.die.net/man/3/pkcs7_sign
+    // https://github.com/openssl/openssl/blob/79c98fc6ccab49f02528e06cc046ac61f841a753/crypto/rsa/rsa_ameth.c#L438
+    digest = EVP_sha256();
+  } else if (!ok) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_NO_DEFAULT_DIGEST);
+    goto err;
+  }
+
+  if ((si = PKCS7_SIGNER_INFO_new()) == NULL ||
+      !PKCS7_SIGNER_INFO_set(si, x509, pkey, digest) ||
```

Comment:
The documentation for [PKCS7_SIGNER_INFO_set](https://github.com/aws/aws-lc/blob/751c814696b245a8e8eefbdb05ab6ac17e7737a8/include/openssl/pkcs7.h#L292-L297) indicates that it bumps the reference count on pkey.
```
|pkey|'s reference count is incremented, but neither |x509|'s nor |dgst|'s is.
```
I'm having a hard time determining whether this might cause a memory leak.  What do you think?

Reply:
We allocate a `PKCS7_SIGNER_INFO` here, then return it to `pkcs7_sign_add_signer`, which in turn returns it to `pkcs7_do_general_sign`, which in turn... drops it on the floor and doesn't do anything with it. I guess that's incidentally OK `p7` will take ownership of `si` in the call to `PKCS7_add_signer` so `si` will get freed when `p7` is freed.

This is confusing and unnecessarily difficult to reason about. I'll change the API of these functions to return ints instead of pointers.